File: flutter_clone/lib/src/wallet_store.dart
class WalletCardData {
  const WalletCardData({
    required this.id,
    required this.holderName,
    required this.number,
    required this.cardType,
    this.balance = 0,
    this.addedAt,
  });

  final String id;
  final String holderName;
  final String number;
  final String cardType;
  final double balance;
  final DateTime? addedAt;

  bool get isTransport => cardType == 'transport';

  String get lastFour =>
      number.length >= 4 ? number.substring(number.length - 4) : number;

  String get maskedNumber => '•••• $lastFour';

  factory WalletCardData.fromJson(Map<String, dynamic> json) {
    return WalletCardData(
      id: json['id'] as String? ?? '',
      holderName: json['holderName'] as String? ?? 'Моя карта',
      number: json['number'] as String? ?? '',
      cardType: json['cardType'] as String? ?? 'bank',
      balance: (json['balance'] as num?)?.toDouble() ?? 0,
      addedAt: DateTime.tryParse(json['addedAt'] as String? ?? ''),
    );
  }

  Map<String, dynamic> toJson() => {
        'id': id,
        'holderName': holderName,
        'number': number,
        'cardType': cardType,
        'balance': balance,
        'addedAt': addedAt?.toIso8601String(),
      };
}

class WalletTransactionData {
  const WalletTransactionData({
    required this.id,
    required this.type,
    required this.source,
    required this.description,
    required this.amount,
    required this.balanceBefore,
    required this.balanceAfter,
    required this.createdAt,
  });

  final String id;
  final String type;
  final String source;
  final String description;
  final double amount;
  final double balanceBefore;
  final double balanceAfter;
  final DateTime createdAt;

  factory WalletTransactionData.fromJson(Map<String, dynamic> json) {
    return WalletTransactionData(
      id: json['id'] as String? ?? '',
      type: json['type'] as String? ?? '',
      source: json['source'] as String? ?? '',
      description: json['description'] as String? ?? '',
      amount: (json['amount'] as num?)?.toDouble() ?? 0,
      balanceBefore: (json['balanceBefore'] as num?)?.toDouble() ?? 0,
      balanceAfter: (json['balanceAfter'] as num?)?.toDouble() ?? 0,
      createdAt: DateTime.tryParse(json['createdAt'] as String? ?? '') ??
          DateTime.now(),
    );
  }
}

class WalletState {
  const WalletState({
    required this.balance,
    required this.cards,
    required this.transactions,
    this.activeCardId,
  });

  final double balance;
  final List<WalletCardData> cards;
  final List<WalletTransactionData> transactions;
  final String? activeCardId;

  const WalletState.empty()
      : balance = 0,
        cards = const [],
        transactions = const [],
        activeCardId = null;

  WalletCardData? get activeCard {
    if (cards.isEmpty) {
      return null;
    }
    for (final card in cards) {
      if (card.id == activeCardId) {
        return card;
      }
    }
    return cards.first;
  }

  WalletState copyWith({
    double? balance,
    List<WalletCardData>? cards,
    List<WalletTransactionData>? transactions,
    String? activeCardId,
    bool clearActiveCard = false,
  }) {
    return WalletState(
      balance: balance ?? this.balance,
      cards: cards ?? this.cards,
      transactions: transactions ?? this.transactions,
      activeCardId: clearActiveCard ? null : (activeCardId ?? this.activeCardId),
    );
  }

  factory WalletState.fromJson(Map<String, dynamic> json) {
    final cards = ((json['cards'] as List<dynamic>? ?? const <dynamic>[])
            .whereType<Map<String, dynamic>>())
        .map(WalletCardData.fromJson)
        .toList();
    final transactions =
        ((json['transactions'] as List<dynamic>? ?? const <dynamic>[])
                .whereType<Map<String, dynamic>>())
            .map(WalletTransactionData.fromJson)
            .toList();

    return WalletState(
      balance: (json['balance'] as num?)?.toDouble() ?? 0,
      cards: cards,
      transactions: transactions,
      activeCardId: json['activeCardId'] as String?,
    );
  }
}

String normalizeCardNumber(String value) {
  return value.replaceAll(RegExp(r'[^0-9A-Za-z]'), '').toUpperCase();
}

String formatBalance(double value) {
  return value.toStringAsFixed(2).replaceAll('.', ',');
}

File: react_native_clone/src/wallet-store.ts
export type WalletCardData = {
  id: string;
  holderName: string;
  number: string;
  cardType: "bank" | "transport";
  balance?: number;
};

export type WalletState = {
  balance: number;
  cards: WalletCardData[];
  activeCardId: string | null;
};

export const emptyWalletState: WalletState = {
  balance: 0,
  cards: [],
  activeCardId: null,
};

export function walletStorageKey(phoneNumber: string) {
  const normalized = phoneNumber.replace(/\D/g, "");
  return `wallet_state_${normalized}`;
}

export function normalizeCardNumber(value: string) {
  return value.replace(/\s+/g, "").toUpperCase();
}

export function maskCardNumber(value: string) {
  const normalized = normalizeCardNumber(value);
  const lastFour = normalized.slice(-4);
  return lastFour ? `•••• ${lastFour}` : "Стандарт";
}

export function formatBalance(value: number) {
  return value.toFixed(2).replace(".", ",");
}

export function hydrateWalletState(raw: unknown): WalletState {
  const state = (raw ?? emptyWalletState) as Partial<WalletState> & {
    cards?: Array<Partial<WalletCardData>>;
  };

  const cards: WalletCardData[] = (state.cards ?? []).map((card) => ({
    id: card.id ?? `${Date.now()}`,
    holderName: card.holderName ?? "Моя карта",
    number: card.number ?? "",
    cardType: card.cardType === "transport" ? "transport" : "bank",
    balance: typeof card.balance === "number" ? card.balance : 0,
  }));

  return {
    balance: typeof state.balance === "number" ? state.balance : 0,
    activeCardId: state.activeCardId ?? cards[0]?.id ?? null,
    cards,
  };
}

export function getActiveCard(walletState: WalletState) {
  return (
    walletState.cards.find((card) => card.id === walletState.activeCardId) ??
    walletState.cards[0] ??
    null
  );
}
